File: ios/AutoskeletonDebugOverlay.swift
import QuartzCore
import UIKit

// Task 3.3 (tasks.md Phase 3) / spec.md REQ-OBS-OVERLAY-1, brief §11: the iOS
// `debugOverlay` — outlines every detected shape with its index, source type, and a
// cache hit/miss badge, so a developer can answer "why was this node not detected"
// by looking for the SHAPE THAT ISN'T THERE (spec.md §2.3's "Debugging a missed
// node" scenario: the absence of an outline is itself the diagnostic signal).
//
// Dev-only, stripped from release: the entire type is wrapped in `#if DEBUG`, a
// true compile-time conditional in Swift (like the C preprocessor, not a runtime
// branch) — in a Release build this type literally does not exist in the compiled
// binary, verified by a dedicated release-configuration build + symbol-table check
// (not expressible as an XCTest case, since test schemes always build Debug; see
// the Phase 3 apply-progress notes for the exact `xcodebuild`/`nm` commands used).
enum AutoskeletonDebugOverlayAvailability {
    static var isAvailable: Bool {
        #if DEBUG
        return true
        #else
        return false
        #endif
    }
}

#if DEBUG
final class AutoskeletonDebugOverlay {
    struct Badge: Equatable {
        let index: Int
        let source: AutoskeletonShapeSource
        let cacheHit: Bool
        let radiusSource: AutoskeletonRadiusSource
    }

    private var outlineLayers: [CAShapeLayer] = []
    private var badgeLayers: [CATextLayer] = []
    private(set) var badges: [Badge] = []

    var shapeCount: Int { outlineLayers.count }

    /// Draws one outline + one badge label per shape (REQ-OBS-OVERLAY-1). Clears
    /// any previously mounted overlay first, so re-mounting on a data update never
    /// leaves stale outlines behind.
    func mount(on surface: UIView, shapes: [AutoskeletonShapeInfo], cacheHit: Bool) {
        clear()
        for (index, shape) in shapes.enumerated() {
            let rect = CGRect(x: shape.x, y: shape.y, width: shape.w, height: shape.h)
            let radius = min(shape.r, min(shape.w, shape.h) / 2)

            let outline = CAShapeLayer()
            outline.path = CGPath(
                roundedRect: rect,
                cornerWidth: max(radius, 0),
                cornerHeight: max(radius, 0),
                transform: nil
            )
            outline.strokeColor = UIColor.systemRed.cgColor
            outline.fillColor = UIColor.clear.cgColor
            outline.lineWidth = 1
            surface.layer.addSublayer(outline)
            outlineLayers.append(outline)

            let badge = CATextLayer()
            badge.string = "\(index) \(shape.source.rawValue) \(cacheHit ? "HIT" : "MISS")"
            badge.fontSize = 9
            badge.foregroundColor = UIColor.systemRed.cgColor
            badge.contentsScale = surface.window?.screen.scale ?? UIScreen.main.scale
            badge.frame = CGRect(x: rect.minX, y: rect.minY, width: rect.width, height: 12)
            surface.layer.addSublayer(badge)
            badgeLayers.append(badge)

            badges.append(Badge(index: index, source: shape.source, cacheHit: cacheHit, radiusSource: shape.radiusSource))
        }
    }

    func clear() {
        outlineLayers.forEach { $0.removeFromSuperlayer() }
        badgeLayers.forEach { $0.removeFromSuperlayer() }
        outlineLayers.removeAll()
        badgeLayers.removeAll()
        badges.removeAll()
    }
}
#endif

File: android/src/main/java/com/autoskeleton/AutoskeletonDebugOverlay.kt
package com.autoskeleton

import android.content.Context
import android.content.pm.ApplicationInfo
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Paint
import android.graphics.RectF
import android.view.View
import android.view.ViewGroup
import kotlin.math.min

// Task 4.5 (tasks.md Phase 4) / spec.md REQ-OBS-OVERLAY-1, plan.md ADR-2: the
// Android `debugOverlay` — outlines every detected shape with its index, source
// type, cache hit/miss badge, AND (ADR-2's mandatory addition on this platform) its
// radius-resolution rung, so a developer can see Android's radius degradation
// directly instead of guessing at it (plan.md: "the debug overlay badges each shape
// with its rung").

class AutoskeletonDebugOverlay internal constructor(context: Context) : View(context) {
    data class Badge(
        val index: Int,
        val source: AutoskeletonShapeSource,
        val cacheHit: Boolean,
        val radiusSource: AutoskeletonRadiusSource,
    )

    private var shapes: List<AutoskeletonShapeInfo> = emptyList()
    private var cacheHit: Boolean = false

    private val outlinePaint = Paint(Paint.ANTI_ALIAS_FLAG).apply {
        style = Paint.Style.STROKE
        strokeWidth = 2f
        color = Color.RED
    }
    private val badgePaint = Paint(Paint.ANTI_ALIAS_FLAG).apply {
        color = Color.RED
        textSize = 24f
    }

    /** Test/telemetry seam: the badges from the most recent [update] call —
     *  REPLACED (never appended to) on every call, so re-mounting on a data update
     *  never leaves a stale badge behind (mirrors iOS's `mount()` "clears any
     *  previously mounted overlay first"). */
    var lastBadges: List<Badge> = emptyList()
        private set

    /** REQ-OBS-OVERLAY-1: one outline + one badge (index, source, cache hit/miss,
     *  AND ADR-2's radius rung) per shape. */
    fun update(shapes: List<AutoskeletonShapeInfo>, cacheHit: Boolean) {
        this.shapes = shapes
        this.cacheHit = cacheHit
        lastBadges = shapes.mapIndexed { index, shape ->
            Badge(index = index, source = shape.source, cacheHit = cacheHit, radiusSource = shape.radiusSource)
        }
        invalidate()
    }

    fun clear() {
        shapes = emptyList()
        lastBadges = emptyList()
        invalidate()
    }

    override fun onDraw(canvas: Canvas) {
        for ((index, shape) in shapes.withIndex()) {
            val radius = min(shape.r, min(shape.w, shape.h) / 2f).coerceAtLeast(0f)
            val rect = RectF(shape.x, shape.y, shape.x + shape.w, shape.y + shape.h)
            canvas.drawRoundRect(rect, radius, radius, outlinePaint)

            val hitMiss = if (cacheHit) "HIT" else "MISS"
            val badgeText = "$index ${shape.source.wireValue} $hitMiss ${shape.radiusSource.wireValue}"
            canvas.drawText(badgeText, shape.x, (shape.y - 4f).coerceAtLeast(badgePaint.textSize), badgePaint)
        }
    }
}

/** Dev-only mount gate. **Stated deviation from iOS's `#if DEBUG` compile-time
 *  strip**: a published Android AAR is a single already-compiled variant, so a
 *  Swift-style build-time strip is not available to a library the way it is to a
 *  CocoaPods pod compiled from source into the app. The standard, portable
 *  mechanism for a library to gate dev-only behavior on the CONSUMING app's own
 *  debug/release build is the runtime `ApplicationInfo.FLAG_DEBUGGABLE` check
 *  below — the same technique Android's own tooling (e.g. `StrictMode`) uses for
 *  the same purpose. */
object AutoskeletonDebugOverlayFactory {
    fun isDebugBuild(context: Context): Boolean =
        (context.applicationInfo.flags and ApplicationInfo.FLAG_DEBUGGABLE) != 0

    /** Mounts a fresh [AutoskeletonDebugOverlay] on [surface] and returns it, or
     *  returns `null` (mounting nothing) when the consuming app is not debuggable. */
    fun createIfDebug(surface: ViewGroup): AutoskeletonDebugOverlay? {
        if (!isDebugBuild(surface.context)) {
            return null
        }
        val overlay = AutoskeletonDebugOverlay(surface.context)
        surface.addView(
            overlay,
            ViewGroup.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.MATCH_PARENT),
        )
        val widthSpec = View.MeasureSpec.makeMeasureSpec(surface.width, View.MeasureSpec.EXACTLY)
        val heightSpec = View.MeasureSpec.makeMeasureSpec(surface.height, View.MeasureSpec.EXACTLY)
        overlay.measure(widthSpec, heightSpec)
        overlay.layout(0, 0, surface.width, surface.height)
        return overlay
    }
}
